### PR TITLE
create stays app

### PR DESCRIPTION
**Pull request by Camelia-Br**
Create a stays app

- Create a Stay model to represent the various stays between an owner and a provider with columns:
      owner: Foreign key to customers.Person
      provider: Foreign key to customers.Provider
      start_date
      end_date
      pets: ManyToMany relationship to customers.Pet
- Create a Review model that stores reviews left by the owner for the provider on a specific Stay
      stay: OneToOneField relationship to Stay
      review
      rating

- Create a StayForm that extends ModelForm and provides the following validations (with tests):
      Stay's start date does not occur after its end date
      Stay's owner is not the same Person as the Stay's Provider's Person. A Provider cannot perform a Stay for themselves

- Create a ReviewForm that extends ModelForm and provides the following validations (with tests):
      Write tests to assert that a Review's rating must be between 1 and 5, inclusive.

[Ticket](https://roverdotcom.atlassian.net/jira/software/c/projects/DEV/boards/402?modal=detail&selectedIssue=DEV-82836)

### CODE REVIEW

File: django_project/settings.py [hunk 1]
```diff
@@ -37,7 +37,9 @@
     'django.contrib.sessions',
     'django.contrib.messages',
     'django.contrib.staticfiles',
+    'django_extensions',
```
Comment (teodorvoicencu) [inline django_project/settings.py:40]: I don't see this anywhere being used. Any reasoning why did you add it ?

File: stays/forms.py [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+
+    def clean(self):
+        cleaned_data = super().clean()
+        owner = cleaned_data.get("owner")
+        provider = cleaned_data.get("provider")
+        start_date = cleaned_data.get("start_date")
+        end_date = cleaned_data.get("end_date")
+        if owner == provider.person:
+            self.add_error('provider', 'You cannot perform a stay for yourself')
+        if end_date < start_date:
```
Comment (teodorvoicencu) [inline stays/forms.py:34]: You should also create a `clean_start_date` and `clean_end_date` such that they are not less than the current date

File: stays/forms.py [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+    class Meta:
+        model = Review
+        fields = ['rating']
+
+    def clean(self):
+        cleaned_data = super().clean()
+        rating = cleaned_data.get("rating")
+
+        if int(rating) > 5:
+            self.add_error('rating', 'You can choose between 1 and 5')
```
Comment (teodorvoicencu) [inline stays/forms.py:60]: You can create a `clean_rating` method that should to this validation

File: stays/tests/forms_tests.py [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+from stays.forms import StayForm, ReviewForm
+from customers.tests.factories import PersonFactory, ProviderFactory, PetFactory
+
+
+class StayFormTests(TestCase):
+    def test_owner_is_the_same_person_as_provider(self):
+        owner = PersonFactory.create()
+        provider = ProviderFactory.create(person=owner)
+        pet = PetFactory.create()
+        data = {"owner": owner, "provider": provider, "start_date": "2022-01-01", "end_date": "2022-02-02", "pets": [pet]}
```
Comment (teodorvoicencu) [inline stays/tests/forms_tests.py]: The date is mostly repeating itself. You can define the setup method where all the necessary objects  used for the tests. Also another method `get_data(self, **kwargs)` should be defined that lets you get the data object you want and makes it easier to change the properties based on the test needs.

File: stays/tests/forms_tests.py [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+        provider = ProviderFactory.create()
+        pet = PetFactory.create()
+        data = {"owner": owner, "provider": provider, "start_date": "2022-03-03", "end_date": "2022-04-04", "pets": [pet]}
+        form = StayForm(data=data)
+        self.assertTrue(form.is_valid())
+
+
+class ReviewFormTest(TestCase):
+    def test_rating_over_5(self):
+        data = {"rating": "6"}
```
Comment (teodorvoicencu) [inline stays/tests/forms_tests.py]: Same comment above applies here as well

File: stays/tests/models_tests.py [hunk 6]
```diff
@@ -0,0 +1,32 @@
+from django.test import TestCase
+from stays.models import Stay, Review
+from .factories import StayFactory, ReviewFactory
+from customers.tests.factories import PersonFactory, PetFactory
+from django.core.exceptions import ValidationError
+
+
+class StayTestCase(TestCase):
+    def test_create_stay(self):
```
Comment (teodorvoicencu) [inline stays/tests/models_tests.py:13]: You can make this more readable by implementing the `setUp` method

File: django_project/settings.py (hunk 1 above)
Comment (Camelia-Br) [reply django_project/settings.py:40]: I added it for shell_plus

File: stays/forms.py [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+        data = self.cleaned_data['start_date']
+        if data and data > now:
+            self.add_error('start_date', 'Pick a valid date')
+        return data
+
+    def clean_end_date(self):
+        now = date.today()
+        data = self.cleaned_data['end_date']
+        if data and data > now:
+            self.add_error('end_date', 'Pick a valid date')
```
Comment (teodorvoicencu) [inline stays/forms.py]: The message for the error is pretty vague.

File: stays/models.py [hunk 8]
```diff
@@ -0,0 +1,28 @@
+from django.db import models
+from customers.models import Person, Provider, Pet
+
+
+class Stay(models.Model):
+    owner = models.ForeignKey(Person, on_delete=models.CASCADE)
+    provider = models.ForeignKey(Provider, on_delete=models.CASCADE)
+    start_date = models.DateField()
+    end_date = models.DateField()
```
Comment (teodorvoicencu) [inline stays/models.py:9]: Shouldn't those fields be required ? Also i would add a default value for start_date

File: stays/models.py [hunk 9]
```diff
@@ -0,0 +1,28 @@
...
+    pets = models.ManyToManyField(Pet)
+
+    def __str__(self):
+        return f"Stay for - {self.owner} - {self.id}"
+
+    def __repr__(self):
+        return f"Stay for - {self.owner} - {self.id}"
```
Comment (teodorvoicencu) [inline stays/models.py]: The goal of `__str__` is to be readable, the goal of `__repr__` is to be unambiguous. I would give more informations about the model on the `__repr__` method and diferentiate it from the `__str__`. You could define something like `"Stay(id=self.id, owner={self.owner}, provider={self.provide}, ...)"` for the `__repr__`

File: stays/tests/forms_tests.py [hunk 10]
```diff
@@ -0,0 +1,64 @@
...
+    def test_start_date_greater_than_end_date(self):
+        data = self.get_data(start_date='2022-02-02', end_date='2022-01-01')
+        form = StayForm(data=data)
+        self.assertFalse(form.is_valid())
+        self.assertEqual(form.errors['end_date'], ['End date should be greater than start date.'])
+
+    def test_end_date_greater_than_start_date(self):
+        data = self.get_data()
+        form = StayForm(data=data)
+        self.assertTrue(form.is_valid())
```
Comment (teodorvoicencu) [inline stays/tests/forms_tests.py:47]: I would also add a test to see that the form is valid if valid data is provided

File: stays/models.py (hunk 8 above)
Comment (Camelia-Br) [reply stays/models.py:9]: DateField() is not required by default?

Comment (teodorvoicencu) [reply stays/models.py:9]: It is, my bad!

File: stays/forms.py [hunk 11]
```diff
@@ -0,0 +1,56 @@
...
+class StayForm(ModelForm):
+    class Meta:
+        model = Stay
+        fields = ['owner', 'provider', 'start_date', 'end_date', 'pets']
+
+    def clean_start_date(self):
+        now = date.today()
+        data = self.cleaned_data['start_date']
+        if data and data > now:
+            self.add_error('start_date', 'Pick a valid date')
```
Comment (teodorvoicencu) [inline stays/forms.py]: The error message is pretty ambiguous

File: stays/tests/forms_tests.py [hunk 12]
```diff
@@ -0,0 +1,71 @@
...
+
+    def test_form_is_valid(self):
+        data = self.get_data()
+        form = StayForm(data=data)
+        self.assertTrue(form.is_valid())
+        form.save()
+        self.assertTrue(Stay.objects.get())
```
Comment (teodorvoicencu) [inline stays/tests/forms_tests.py]:
```suggestion
        stay = form.save()
        self.assertIsNotNone(stay)
```

File: stays/forms.py (hunk 11 above)
Comment (Camelia-Br) [reply stays/forms.py]: Start date should not be from the past???

**Review by teodorvoicencu — APPROVED** [review]
